beacon-chain/sync/pending_blocks_queue.go

```diff
@@ -132,85 +116,123 @@ func (s *Service) processPendingBlocks(ctx context.Context) error {
 				continue
 			}
 
+			// Request parent block if not in the pending queue and not in the database.
 			parentRoot := b.Block().ParentRoot()
-			parentInDb := s.cfg.beaconDB.HasBlock(ctx, parentRoot)
-			hasPeer := len(pids) != 0
-
-			// Only request for missing parent block if it's not in beaconDB, not in pending cache
-			// and has peer in the peer list.
-			if !inPendingQueue && !parentInDb && hasPeer {
-				log.WithFields(logrus.Fields{
-					"currentSlot": b.Block().Slot(),
-					"parentRoot":  hex.EncodeToString(bytesutil.Trunc(parentRoot[:])),
-				}).Debug("Requesting parent block")
-				parentRoots = append(parentRoots, b.Block().ParentRoot())
-
-				span.End()
+			inPendingQueue := s.isBlockInQueue(b.Block().ParentRoot())
+			if s.shouldRequestParentBlock(ctx, inPendingQueue, parentRoot) {
+				log.WithFields(logrus.Fields{"currentSlot": b.Block().Slot(), "parentRoot": hex.EncodeToString(parentRoot[:])}).Debug("Requesting parent block")
+				parentRoots = append(parentRoots, parentRoot)
 				continue
 			}
-
-			if !parentInDb {
-				span.End()
+			if !s.cfg.beaconDB.HasBlock(ctx, parentRoot) {
```

Comment:
hmm so performing this check in the helper causes you to hit the db twice. I would reconsider that aspect.

Reply:
cached the result using `parentBlockInDB` in [0dfcbfd](https://github.com/prysmaticlabs/prysm/pull/13026/commits/0dfcbfd95d0c908729ae571002c5f6dbbe06e105)